Mathlib/Control/Monad/Freer.lean

```diff
@@ -0,0 +1,290 @@
+    case impure X Fx k ih =>
+      simp [Freer.map, Seq.seq, bindFree, Functor.map, SeqRight.seqRight] at ih ⊢
+      apply funext; exact ih
+  pure_seq f x := by
+    simp [Seq.seq, Functor.map, pure, bindFree]
+
+/-! ### State Monad via `Freer` -/
+
+/-- Type constructor for state operations. -/
+inductive StateF (σ : Type u) : Type u → Type _ where
+  | get : StateF σ σ                   -- get the current state
+  | set : σ → StateF σ PUnit           -- set the state
+
+/-- State monad via the `Freer` monad. -/
+abbrev FreerState (σ : Type u) := Freer (StateF σ)
+
+namespace FreerState
+
+instance {σ : Type u} : Monad (FreerState σ) := inferInstance
+instance {σ : Type u} : LawfulMonad (FreerState σ) := inferInstance
+
+instance {σ : Type u} : MonadStateOf σ (FreerState σ) where
+  get := Freer.impure σ (StateF.get) Freer.pure
+  set newState := Freer.impure PUnit (StateF.set newState) (fun _ => Freer.pure PUnit.unit)
+  modifyGet f :=
+    Freer.impure σ StateF.get (fun s =>
+      let (a, s') := f s
+      Freer.impure PUnit (StateF.set s') (fun _ => Freer.pure a))
+
+instance {σ : Type u} : MonadState σ (FreerState σ) := inferInstance
+
+/-- Run a state computation, returning both the result and final state. -/
+def runState {σ : Type u} {α : Type v} (comp : FreerState σ α) (s₀ : σ) : α × σ :=
+  match comp with
+  | .pure a => (a, s₀)
+  | .impure _ StateF.get k     => runState (k s₀) s₀
+  | .impure _ (StateF.set s') k => runState (k PUnit.unit) s'
+
+/-- Run a state computation, returning only the result. -/
+def evalState {σ : Type u} {α : Type v} (c : FreerState σ α) (s₀ : σ) : α :=
+  (runState c s₀).1
+
+/-- Run a state computation, returning only the final state. -/
+def execState {σ : Type u} {α : Type v} (c : FreerState σ α) (s₀ : σ) : σ :=
+  (runState c s₀).2
+
+end FreerState
+
+/-! ### Writer Monad via `Freer` -/
+
+/-- Type constructor for writer operations. -/
+inductive WriterF (w : Type u) : Type v → Type _ where
+  /-- Append a value to the log. -/
+  | tell : w → WriterF w PUnit
+
+/-- Writer monad via the `Freer` monad. -/
+abbrev FreerWriter (w : Type u) := Freer (WriterF w)
+
+namespace FreerWriter
+
+instance {w : Type u} : Monad (FreerWriter w) := inferInstance
+instance {w : Type u} : LawfulMonad (FreerWriter w) := inferInstance
+
+/-- Append to the log. -/
+def tell {w : Type u} (log : w) : FreerWriter w PUnit :=
+  Freer.impure PUnit (WriterF.tell log) Freer.pure
+
+/-- Run a writer computation, returning the result and log. -/
+def runWriter {w : Type u} {α : Type v} [AddMonoid w] (computation : FreerWriter w α) : α × w :=
+  match computation with
+  | Freer.pure a => (a, 0)
+  | Freer.impure _ (WriterF.tell log) cont =>
+      let (result, accLog) := runWriter (cont PUnit.unit)
+      (result, log + accLog)
+
+/-- Capture the output produced by a computation. -/
+def listen {w : Type u} {α : Type v} [AddMonoid w] (comp : FreerWriter w α) :
+FreerWriter w (α × w) := Freer.pure (runWriter comp)
```

Comment:
My understanding here is that you need to re-emit what was previously logged after capturing it